### ISSUE TITLE
Increase the kernel -> fuse write size

### ISSUE THREAD

**Issue by jacobsa** [GoogleCloudPlatform/gcsfuse#27]
As seen in #22, enabling fuse "big writes" ups the kernel -> gcsfuse write atom from 4 KiB to 32 KiB, which makes a performance difference.

Things to do:

* Research, document, and support this in jacobsa/gcsfuse. Probably enabled by default (with a callout in the docs).

* Can Linux be made to send even larger writes? Setting `InitResponse.MaxWrite` didn't appear to work.

* Google around for "fuse write performance" and see if there's anything else obvious to be done.

**Comment by jacobsa** [comment]
[`FUSE_BIG_WRITES`](https://github.com/torvalds/linux/blob/d7afaec0b564f0609e116f562983b8e72fc3e9c9/include/uapi/linux/fuse.h#L221) was added in torvalds/linux@78bb6cb9a890d3d50ca3b02fce9223d3e734ab9b.

It is used in `inode.c` [to initialize](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/inode.c#L870-L871) the connection field `big_writes`. That field is used in only [one place](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/file.c#L1065-L1066), in `file.c`, to break early from a loop in `fuse_fill_write_pages` after one page that would otherwise write up to `max_write`/`max_pages` pages.

**Comment by jacobsa** [comment]
Investigation:

* `max_pages` is [set](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/dev.c#L51) by `fuse_request_init`.

* That is [called](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/dev.c#L77) by `__fuse_request_alloc`.

* Which is [called](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/dev.c#L172) by `__fuse_get_req`.

* Which is [called](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/file.c#L1102) by `fuse_perform_write`.

* Who gets the `nrpages` argument [from](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/file.c#L1100) `fuse_wr_pages`.

* Who [computes](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/file.c#L1073-L1079) it as the minimum of the pages requested and `FUSE_MAX_PAGES_PER_REQ`.

* Which is [defined](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/fuse_i.h#L26-L27) to be 32.

So I think we should be able to drive this up to 128 KiB. Maybe the problem is that `MaxWrite` is supposed to be a count of pages.

**Comment by jacobsa** [comment]
No, `max_write` appears to be used [here](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/inode.c#L895-L896) and [here](https://github.com/torvalds/linux/blob/b83ae6d421435c6204150300f1c25bfbd39cd62b/fs/fuse/file.c#L1028) as a count of bytes. Hmm.

**Comment by jacobsa** [comment]
On OS X, `MemFSTest.LargeFile` receives 1 MiB writes, and on Linux it's 4 KiB. Yeesh.

**Comment by jacobsa** [comment]
My test setup on GCE must have been mistaken. If I set `MaxWrite` high and return `InitBigWrites`, I get writes of 128 KiB as expected. Hooray.

**Comment by jdlourenco** [comment]
Hi @jacobsa, I'm having trouble increasing the size of the FUSE big_writes mode in order to perform writes of more than 64 kb at a time using the C libfuse library, version 2.9.7.

I was able to increase the amount of data handled on each write to 64 kb by specifying the "-o big_writes" option.

I then tried to use the "-o max_write=131072" option to set the write size at 128 kb but I keep seeing writes of only 64 kb.

Were you able to set this value to 128kb, what variable or option must I use/set.

Thanks in advance.
